File: mockserver-client-dotnet/src/MockServer.Client/Models/Llm/LlmFailoverBuilder.cs
using MockServer.Client.Models;

namespace MockServer.Client.Llm;

/// <summary>
/// Builder for provider failover/retry scenarios (mirrors
/// <c>org.mockserver.client.LlmFailoverBuilder</c>).
///
/// Produces an ordered list of expectations: limited-<c>times</c> failure
/// expectations first (consumed before the success), then a single
/// unlimited-<c>times</c> success expectation carrying an <c>httpLlmResponse</c>.
/// Consecutive identical failures are coalesced into one expectation with
/// <c>times.remainingTimes = count</c>.
/// </summary>
public sealed class LlmFailoverBuilder
{
    private sealed class FailureSpec
    {
        public FailureSpec(int statusCode, string? errorBody)
        {
            StatusCode = statusCode;
            ErrorBody = errorBody;
        }

        public int StatusCode { get; }
        public string? ErrorBody { get; }
    }

    private string? _path;
    private string? _provider;
    private string? _model;
    private readonly List<FailureSpec> _failures = new();
    private Completion? _successCompletion;

    /// <summary>Entry point mirroring <c>LlmFailoverBuilder.llmFailover()</c>.</summary>
    public static LlmFailoverBuilder LlmFailover() => new();

    public LlmFailoverBuilder WithPath(string path)
    {
        _path = path;
        return this;
    }

    public LlmFailoverBuilder WithProvider(string provider)
    {
        _provider = provider;
        return this;
    }

    public LlmFailoverBuilder WithModel(string model)
    {
        _model = model;
        return this;
    }

    /// <summary>Adds one failure attempt with the default error body for the status code.</summary>
    public LlmFailoverBuilder FailWith(int statusCode)
    {
        ValidateStatusCode(statusCode);
        _failures.Add(new FailureSpec(statusCode, null));
        return this;
    }

    /// <summary>Adds one failure attempt with a custom JSON error body.</summary>
    public LlmFailoverBuilder FailWith(int statusCode, string errorBody)
    {
        ValidateStatusCode(statusCode);
        _failures.Add(new FailureSpec(statusCode, errorBody));
        return this;
    }

    /// <summary>Adds <paramref name="count"/> failure attempts with the default error body.</summary>
    public LlmFailoverBuilder FailWith(int statusCode, int count)
    {
        ValidateStatusCode(statusCode);
        if (count < 1)
            throw new ArgumentException($"count must be >= 1, got {count}", nameof(count));
        for (var i = 0; i < count; i++)
            _failures.Add(new FailureSpec(statusCode, null));
        return this;
    }

    public LlmFailoverBuilder ThenRespondWith(Completion completion)
    {
        _successCompletion = completion;
        return this;
    }

    public int GetFailureCount() => _failures.Count;

    public List<Expectation> Build()
    {
        if (string.IsNullOrEmpty(_path))
            throw new InvalidOperationException("Path must be set");
        if (string.IsNullOrEmpty(_provider))
            throw new InvalidOperationException("Provider must be set");
        if (_failures.Count == 0)
            throw new InvalidOperationException("At least one failure must be defined");
        if (_successCompletion == null)
            throw new InvalidOperationException("Success completion must be set via ThenRespondWith()");

        var expectations = new List<Expectation>();
        foreach (var (statusCode, errorBody, count) in Coalesce())
        {
            var body = errorBody ?? DefaultErrorBody(statusCode);
            expectations.Add(new Expectation
            {
                HttpRequest = new HttpRequest { Method = "POST", Path = _path },
                Times = Times.Exactly(count),
                TimeToLive = TimeToLive.UnlimitedTtl(),
                HttpResponse = new HttpResponse
                {
                    StatusCode = statusCode,
                    Headers = new Dictionary<string, List<string>>
                    {
                        ["Content-Type"] = new() { "application/json" }
                    },
                    Body = body
                }
            });
        }

        expectations.Add(new Expectation
        {
            HttpRequest = new HttpRequest { Method = "POST", Path = _path },
            Times = Times.Unlimited(),
            TimeToLive = TimeToLive.UnlimitedTtl(),
            HttpLlmResponse = new HttpLlmResponse
            {
                Provider = _provider,
                Model = _model,
                Completion = _successCompletion
            }
        });

        return expectations;
    }

    public List<Expectation> ApplyTo(MockServerClient client) => client.Upsert(Build().ToArray());

    private static void ValidateStatusCode(int statusCode)
    {
        if (statusCode < 100 || statusCode > 599)
            throw new ArgumentException($"statusCode must be between 100 and 599, got {statusCode}", nameof(statusCode));
    }

    private IEnumerable<(int StatusCode, string? ErrorBody, int Count)> Coalesce()
    {
        var result = new List<(int StatusCode, string? ErrorBody, int Count)>();
        foreach (var spec in _failures)
        {
            if (result.Count > 0)
            {
                var last = result[result.Count - 1];
                if (last.StatusCode == spec.StatusCode && last.ErrorBody == spec.ErrorBody)
                {
                    result[result.Count - 1] = (last.StatusCode, last.ErrorBody, last.Count + 1);
                    continue;
                }
            }
            result.Add((spec.StatusCode, spec.ErrorBody, 1));
        }
        return result;
    }

    /// <summary>
    /// Default JSON error body for a status code (exposed for testing / parity).
    /// Shape: <c>{"error":{"type":&lt;type&gt;,"message":&lt;message&gt;}}</c>.
    /// </summary>
    public static string DefaultErrorBody(int statusCode)
    {
        var (type, message) = statusCode switch
        {
            429 => ("rate_limit_error", "Rate limit exceeded. Please retry after a brief wait."),
            500 => ("internal_server_error", "An internal error occurred. Please retry your request."),
            502 => ("bad_gateway", "Bad gateway. The upstream server returned an invalid response."),
            503 => ("service_unavailable", "The service is temporarily overloaded. Please retry later."),
            _ => ("error", $"Request failed with status {statusCode}")
        };
        return "{\"error\":{\"type\":\"" + type + "\",\"message\":\"" + message + "\"}}";
    }
}

File: mockserver-client-php/src/Llm/LlmFailoverBuilder.php
<?php

declare(strict_types=1);

namespace MockServer\Llm;

use InvalidArgumentException;
use MockServer\Expectation;
use MockServer\MockServerClient;

/**
 * Builder for provider failover/retry scenarios
 * (mirrors org.mockserver.client.LlmFailoverBuilder).
 *
 * Produces an ordered list of expectations: failure expectations (limited
 * {@code times}) first — consumed before — a single success expectation with
 * unlimited {@code times}. Consecutive identical failures are coalesced into
 * one expectation with {@code times.remainingTimes = count}.
 *
 * @example
 *   LlmFailoverBuilder::llmFailover()
 *       ->withPath('/v1/chat/completions')
 *       ->withProvider(Provider::OPENAI)
 *       ->failWith(429)
 *       ->failWith(503, 2)
 *       ->thenRespondWith(Completion::completion()->withText('OK'))
 *       ->applyTo($client);
 */
class LlmFailoverBuilder
{
    private ?string $path = null;
    private ?string $provider = null;
    private ?string $model = null;
    /** @var list<array{statusCode: int, errorBody: string|null}> */
    private array $failures = [];
    private ?Completion $successCompletion = null;

    /**
     * Entry point mirroring {@code LlmFailoverBuilder.llmFailover()}.
     */
    public static function llmFailover(): self
    {
        return new self();
    }

    public function withPath(string $path): self
    {
        $this->path = $path;
        return $this;
    }

    public function withProvider(string $provider): self
    {
        $this->provider = $provider;
        return $this;
    }

    public function withModel(string $model): self
    {
        $this->model = $model;
        return $this;
    }

    /**
     * Add one (or {@code count}) failure attempt(s) with the given status.
     *
     * Mirrors the three Java overloads:
     *   failWith(status)
     *   failWith(status, errorBody)   — custom JSON error body
     *   failWith(status, count)       — repeat the default body N times
     *
     * @param string|int|null $errorBodyOrCount A non-empty string is a custom
     *                                           error body; an int is a repeat count.
     */
    public function failWith(int $statusCode, string|int|null $errorBodyOrCount = null): self
    {
        if ($statusCode < 100 || $statusCode > 599) {
            throw new InvalidArgumentException("statusCode must be between 100 and 599, got {$statusCode}");
        }

        if (is_int($errorBodyOrCount)) {
            if ($errorBodyOrCount < 1) {
                throw new InvalidArgumentException("count must be >= 1, got {$errorBodyOrCount}");
            }
            for ($i = 0; $i < $errorBodyOrCount; $i++) {
                $this->failures[] = ['statusCode' => $statusCode, 'errorBody' => null];
            }
        } else {
            $this->failures[] = ['statusCode' => $statusCode, 'errorBody' => $errorBodyOrCount];
        }

        return $this;
    }

    public function thenRespondWith(Completion $completion): self
    {
        $this->successCompletion = $completion;
        return $this;
    }

    public function getFailureCount(): int
    {
        return count($this->failures);
    }

    /**
     * Default JSON error body for a status code, mirroring the Java helper.
     */
    public static function defaultErrorBody(int $statusCode): string
    {
        switch ($statusCode) {
            case 429:
                $type = 'rate_limit_error';
                $message = 'Rate limit exceeded. Please retry after a brief wait.';
                break;
            case 500:
                $type = 'internal_server_error';
                $message = 'An internal error occurred. Please retry your request.';
                break;
            case 502:
                $type = 'bad_gateway';
                $message = 'Bad gateway. The upstream server returned an invalid response.';
                break;
            case 503:
                $type = 'service_unavailable';
                $message = 'The service is temporarily overloaded. Please retry later.';
                break;
            default:
                $type = 'error';
                $message = 'Request failed with status ' . $statusCode;
                break;
        }

        return json_encode(
            ['error' => ['type' => $type, 'message' => $message]],
            JSON_THROW_ON_ERROR
        );
    }

    /**
     * Coalesce consecutive identical failures.
     *
     * @return list<array{statusCode: int, errorBody: string|null, count: int}>
     */
    private function coalesce(): array
    {
        $result = [];
        foreach ($this->failures as $spec) {
            $last = $result[count($result) - 1] ?? null;
            if ($last !== null
                && $last['statusCode'] === $spec['statusCode']
                && $last['errorBody'] === $spec['errorBody']) {
                $result[count($result) - 1]['count']++;
                continue;
            }
            $result[] = ['statusCode' => $spec['statusCode'], 'errorBody' => $spec['errorBody'], 'count' => 1];
        }
        return $result;
    }

    /**
     * @return list<Expectation>
     */
    public function build(): array
    {
        if ($this->path === null) {
            throw new InvalidArgumentException('Path must be set');
        }
        if ($this->provider === null) {
            throw new InvalidArgumentException('Provider must be set');
        }
        if (count($this->failures) === 0) {
            throw new InvalidArgumentException('At least one failure must be defined');
        }
        if ($this->successCompletion === null) {
            throw new InvalidArgumentException('Success completion must be set via thenRespondWith()');
        }

        $expectations = [];
        foreach ($this->coalesce() as $cf) {
            $body = $cf['errorBody'] ?? self::defaultErrorBody($cf['statusCode']);
            $expectations[] = Expectation::fromArray([
                'httpRequest' => ['method' => 'POST', 'path' => $this->path],
                'times' => ['remainingTimes' => $cf['count'], 'unlimited' => false],
                'timeToLive' => ['unlimited' => true],
                'httpResponse' => [
                    'statusCode' => $cf['statusCode'],
                    'headers' => [['name' => 'Content-Type', 'values' => ['application/json']]],
                    'body' => $body,
                ],
            ]);
        }

        $successAction = HttpLlmResponse::llmResponse()
            ->withProvider($this->provider)
            ->withCompletion($this->successCompletion);
        if ($this->model !== null) {
            $successAction->withModel($this->model);
        }

        $expectations[] = Expectation::fromArray([
            'httpRequest' => ['method' => 'POST', 'path' => $this->path],
            'times' => ['remainingTimes' => 0, 'unlimited' => true],
            'timeToLive' => ['unlimited' => true],
            'httpLlmResponse' => $successAction->toArray(),
        ]);

        return $expectations;
    }

    /**
     * @return array<mixed>
     */
    public function applyTo(MockServerClient $client): array
    {
        $results = [];
        foreach ($this->build() as $expectation) {
            $results[] = $client->upsertExpectation($expectation);
        }
        return $results;
    }
}
